Google.Api.Gax.Grpc/HeaderParameterExtractor.cs

```diff
@@ -13,54 +13,89 @@
-        /// <param name="selector"></param>
-        public HeaderParameterExtractor(string paramName, Regex regex, Func<TRequest, string> selector)
+        public HeaderParameterExtractor()
         {
-            _extractions = new List<ParamExtraction> { new ParamExtraction(paramName, regex, selector) };
+            _extractions = new List<ParamExtraction>();
         }
 
         private HeaderParameterExtractor(IEnumerable<ParamExtraction> extractions)
         {
             _extractions = new List<ParamExtraction>(extractions);
+
+            // It is important that we Reverse the parameters within the group.
+            // The extractions follow the `last successfully matched wins` rule for
+            // conflict resolution when there are multiple extractions for the same parameter name.
+            // (see `google/api/routing.proto` for further details)
+            // But we would prefer to stop evaluating on the first successfully matched extraction,
+            // thus the reverse.
+            _groupings = _extractions.GroupBy(pe => pe.ParamName)
+                .ToDictionary(paramGroup => paramGroup.Key,
+                    paramGroup => paramGroup.Reverse());
         }
 
         /// <summary>
-        /// 
+        /// Adds a new header parameter extraction instruction.
+        /// The extractions follow the `last successfully matched wins` rule for
+        /// conflict resolution when there are multiple extractions for the same parameter name.
+        /// (see `google/api/routing.proto` for further details)
         /// </summary>
-        /// <param name="paramName"></param>
-        /// <param name="regex"></param>
-        /// <param name="selector"></param>
+        /// <param name="paramName">The name of the parameter in the routing header.</param>
+        /// <param name="regexStr">The regular expression (in the string form) used to extract the value of the parameter.
+        /// Should have exactly one named capturing group.</param>
+        /// <param name="selector">A function to call on each request, to determine the string to extract the header value from.
+        /// The parameter must not be null, but may return null.</param>
         /// <returns></returns>
-        public HeaderParameterExtractor<TRequest> Add(string paramName, Regex regex, Func<TRequest, string> selector)
+        public HeaderParameterExtractor<TRequest> WithParameter(string paramName, string regexStr, Func<TRequest, string> selector)
         {
+            GaxPreconditions.CheckNotNullOrEmpty(paramName, nameof(paramName));
+            GaxPreconditions.CheckNotNull(regexStr, nameof(regexStr));
+            GaxPreconditions.CheckNotNull(selector, nameof(selector));
+
+            var regex = new Regex(regexStr);
+
+            // All regexes have a capturing group named `0` that captures the whole regex
+            if (regex.GetGroupNames().Length != 2)
+            {
+                var errMsg =
```

Comment:
Nit: unwrap, or just inline it into the throw statement. (I don't think it's worth having a separate local variable.)